test/index.js

```diff
@@ -1463,6 +1463,139 @@ describe('Nightmare', function () {
       didSubscribe.should.be.false;
     });
 
+    it('should subscribe to frames when requested necessary', function(done) {
+      var didSubscribe = false;
+      var didUnsubscribe = false;
+      var FrameManager = require('../lib/frame-manager.js');
+      var fn;
+      var manager = FrameManager({
+        webContents: {
+          debugger: {
+            isAttached: function() { return true; },
+            sendCommand: function(command) { if (command === 'DOM.highlightRect') { fn('mock-data'); }}
+          },
+          beginFrameSubscription: function(_fn) { didSubscribe = true; fn = _fn; },
+          endFrameSubscription: function() { didUnsubscribe = true; },
+          executeJavaScript: function() {}
+        }
+      });
```

Comment:
Oh, that’s not what I was trying to suggest at all. I was thinking you might have something roughly like…

```js
function MockFrameManagerWindow (options) {
  options = Object.assign({
    debuggerAlreadyAttached: false,
    debuggerTriggersFrames: true
  }, options);

  var self = this;
  var _debuggerAttached = !!options.debuggerAlreadyAttached;
  var _frameCallback = function() {};

  self.results = {
    subscribedCount: 0,
    unsubscribedCount: 0,
    currentSubscriptionCount: 0
  };

  self.beginFrameSubscription = function (callback) {
    _frameCallback = callback;
    self.results.subscribedCount++;
    self.results.currentSubscriptionCount++;
  };

  self.endFrameSubscription = function () {
    _frameCallback = function() {};
    self.results.unsubscribedCount++;
    self.results.currentSubscriptionCount--;
  };

  self.debugger = {
    isAttached: function () { return _debuggerAttached; },
    attach: function () {
      asset(!_debuggerAttached, 'Debugger is already attached');
      _debuggerAttached = true;
    },
    detach: function () {
      _debuggerAttached = false;
    },
    sendCommand: function (command, callback) {
      if (options.debuggerTriggersFrames && command === 'DOM.highlightRect') {
        setTimeout(_frameCallback, 0, 'mock_data');
      }
      if (callback) {
        setTimeout(callback, 0);
      }
    }
  };
}
```

If you wanted to do less work here, you could use [Sinon](http://sinonjs.org) to make all the functions; it automatically keeps tracked of whether they were called and how often.

Anyway, instead of using local closure variables, you’d just test `mockInstance.results.whatever` to find out whether things were called or not.